src/Subscriber/PaymentSubscriber.php

```diff
@@ -199,9 +205,29 @@ public static function getSubscribedEvents(): array
      */
     public function onContextTokenUpdate(SalesChannelContextSwitchEvent $event)
     {
-        // Save (or clear) state.data if payment method is selected/updated
+        // Clear state.data if payment method is updated
         if ($event->getRequestDataBag()->has('paymentMethodId')) {
-            $this->saveStateData($event);
+            $this->removeCurrentStateData($event);
+        }
+
+        // Save state data, only if Adyen payment method is selected
+        if ($event->getRequestDataBag()->get('adyenStateData')) {
+            // Use payment method selected in the same request if available, otherwise get payment method from context
+            $paymentMethodId = $event->getRequestDataBag()->get('paymentMethodId')
+                ?? $event->getSalesChannelContext()->getPaymentMethod()->getId();
+            /** @var PaymentMethodEntity $paymentMethod */
+            $paymentMethod = $this->paymentMethodRepository->search(
+                (new Criteria())
+                    ->addFilter(new EqualsFilter('id', $paymentMethodId)),
+                $event->getContext()
+            )->first();
+            if ($paymentMethod->getPluginId() === $this->adyenPluginId) {
+                $this->saveStateData($event, $paymentMethod);
+            } else {
+                $this->logger->error('Please select an Adyen payment method to save payment state data.');
```

Comment:
Since this is a log message maybe this is better?
```suggestion
                $this->logger->error('No Adyen payment method selected, skipping state data save.');
```

Reply:
You're right